src/Network/Ethereum/Web3/Provider.hs

```diff
@@ -80,7 +84,13 @@ runWeb3With manager provider = runWeb3With' manager provider noopLogger
 
 runWeb3With' :: MonadIO m => Manager -> Provider -> Web3Logger -> Web3 a -> m (Either Web3Error a)
 runWeb3With' manager provider w3logger =
-    liftIO . try .  flip runReaderT (provider, manager, w3logger) . unWeb3
+    liftIO . try . unifyErrs . flip runReaderT (provider, manager, w3logger) . unWeb3
+    where
+        unifyErrs :: IO a -> IO a
+        unifyErrs action = action
+            `catch` \case
+                ParsingException s -> throwM $ ParserFail s
+                CallException err -> throwM . JsonRpcFail $ show err
```

Comment:
This block is the meat of this change. Everything else is just a shuffling of definitions to deal with dependency graphs.